playground/infrastructure/datastore_client.py

```diff
@@ -0,0 +1,202 @@
+        Args:
+            :param sdk: sdk from parameters
+            :param examples_from_rep: examples from the repository for saving to the Cloud Datastore
+        """
+        # initialise data
+        snippets = []
+        examples = []
+        pc_objects = []
+        files = []
+        updated_example_ids = []
+        now = datetime.today()
+
+        # retrieve the last schema version
+        actual_schema_version_key = self._get_actual_schema_version_key()
+
+        # retrieve all example keys before updating
+        examples_ids_before_updating = self._get_all_examples(sdk)
+
+        # loop through every example to save them to the Cloud Datastore
+        with self._datastore_client.transaction():
+            for example in tqdm(examples_from_rep):
+                sdk_key = self._get_key(DatastoreProps.SDK_KIND, Sdk.Name(example.sdk))
+                example_id = f"{Sdk.Name(example.sdk)}{config.DatastoreProps.KEY_NAME_DELIMITER}{example.name}"
+                updated_example_ids.append(example_id)
+                self._to_example_entities(example, example_id, sdk_key, actual_schema_version_key, examples)
+                self._to_snippet_entities(example, example_id, sdk_key, now, actual_schema_version_key, snippets)
+                self._to_pc_object_entities(example, example_id, pc_objects)
+                self._to_file_entities(example, example_id, files)
+
+            self._datastore_client.put_multi(examples)
+            self._datastore_client.put_multi(snippets)
+            self._datastore_client.put_multi(pc_objects)
+            self._datastore_client.put_multi(files)
+
+            # delete examples from the Cloud Datastore that are not in the repository
+            examples_ids_for_removing = list(filter(lambda key: key not in updated_example_ids, examples_ids_before_updating))
+            if len(examples_ids_for_removing) != 0:
+                logging.info("Start of deleting extra playground examples ...")
+                examples_keys_for_removing = list(map(lambda ex_id: self._get_key(DatastoreProps.EXAMPLE_KIND, ex_id), examples_ids_for_removing))
+                snippets_keys_for_removing = list(map(lambda ex_id: self._get_key(DatastoreProps.SNIPPET_KIND, ex_id), examples_ids_for_removing))
+                file_keys_for_removing = list(map(lambda ex_id: self._get_key(DatastoreProps.FILED_KIND, f"{ex_id}{config.DatastoreProps.KEY_NAME_DELIMITER}{0}"), examples_ids_for_removing))
+                pc_objs_keys_for_removing = []
+                for example_id_item in examples_ids_for_removing:
+                    for example_type in [PrecompiledExample.GRAPH_EXTENSION.upper(), PrecompiledExample.OUTPUT_EXTENSION.upper(), PrecompiledExample.LOG_EXTENSION.upper()]:
+                        pc_objs_keys_for_removing.append(self._get_key(DatastoreProps.PRECOMPILED_OBJECT_KIND, f"{example_id_item}{config.DatastoreProps.KEY_NAME_DELIMITER}{example_type}"))
```

Comment:
maybe we should add auto-formatting for this code base? so that it's consistent with the rest of the Beam code base in Python (can be done as follow-up)

Reply:
make sense, yep, i can do that in the next PR
i'd like to clarify this point because i have a little experience with Python. @pabloem Do you mean use this utility for auto-formatting: https://pypi.org/project/autopep8/ ?